### ISSUE TITLE
Allow widget: select for type: object, change jsonName and jsonValue to support jsonPointers 

### ISSUE THREAD

**Issue by Salmu96** [dashjoin/json-schema-form#82]
Hello everyone ^ _ ^!
I would like to know if it is possible, given a certain json like the following: https://jsonplaceholder.typicode.com/posts to select only SOME of its fields, a subset.

For example, if I did this: https://stackblitz.com/edit/dashjoin-irugsc?file=src%2Fapp%2Fapp.component.ts I would get all json's fields both as a view of choice for the user in the dropdown menu and as a final value in json printed below, while I would like it to be displayed as jsonName eg "title" and "id", while as jsonValue I would like "body" and "userId" to be printed below as json output value.
Do you know if this is possible? Thanks a lot in advance ^^

**Comment by aeberhart** [comment]
Not sure I understand 100% but it seems like "order" is what you need:
https://dashjoin.github.io/#/example/order
You can change the order, omit fields and also define subgroups.

**Comment by Salmu96** [comment]
> Not sure I understand 100% but it seems like "order" is what you need:
> https://dashjoin.github.io/#/example/order
> You can change the order, omit fields and also define subgroups.

I'm sorry @aeberhart , I think I was not very understandable in my first message, now I try to explain myself better.

Let's Consider this example now: https://stackblitz.com/edit/dashjoin-czvhan?file=src%2Fapp%2Fapp.component.ts
I would like to see in the dropdown menu the value of the property "name" of the json while I see "[object Object]" is this a bug or am I doing something wrong?

![exampleDASHJOIN](https://user-images.githubusercontent.com/72612398/96847791-e44c9f80-1453-11eb-8caa-aef75a3c72a5.PNG)

Thank you in advance ^^

**Comment by aeberhart** [comment]
Let's have a look at the name property definition:

```
{ 
  type: "string", 
  "widget": "select", 
  description: "Please enter your name", 
  choicesUrl: "https://raw.githubusercontent.com/dashjoin/json-schema-form/master/src/assets/autocomplete-complex.json",
  jsonPointer: "/result", 
  jsonName: "name", 
  jsonValue: "", 
  choicesVerb: "GET" 
},
```

choicesUrl and choicesVerb load this JSON structure:

```
{
    "result": [
        {
            "name": "China",
            "url": "https://en.wikipedia.org/wiki/Demographics_of_China"
        },
       ...
```

Then, jsonPointer is applied yielding this:

```
[
   {
       "name": "China",
       "url": "https://en.wikipedia.org/wiki/Demographics_of_China"
   },
  ...
```

jsonName and jsonValue tell the system which fields to use as a display name / as a select value. So if you set jsonValue: "url" the dropdown shows the country names and if you select them, their wikipedia url is added to the result.

**Comment by Salmu96** [comment]
> Let's have a look at the name property definition:
> 
> ```
> { 
>   type: "string", 
>   "widget": "select", 
>   description: "Please enter your name", 
>   choicesUrl: "https://raw.githubusercontent.com/dashjoin/json-schema-form/master/src/assets/autocomplete-complex.json",
>   jsonPointer: "/result", 
>   jsonName: "name", 
>   jsonValue: "", 
>   choicesVerb: "GET" 
> },
> ```
> 
> choicesUrl and choicesVerb load this JSON structure:
> 
> ```
> {
>     "result": [
>         {
>             "name": "China",
>             "url": "https://en.wikipedia.org/wiki/Demographics_of_China"
>         },
>        ...
> ```
> 
> Then, jsonPointer is applied yielding this:
> 
> ```
> [
>    {
>        "name": "China",
>        "url": "https://en.wikipedia.org/wiki/Demographics_of_China"
>    },
>   ...
> ```
> 
> jsonName and jsonValue tell the system which fields to use as a display name / as a select value. So if you set jsonValue: "url" the dropdown shows the country names and if you select them, their wikipedia url is added to the result.

Hi @aeberhart,
Many thanks for your assistance.
I agree with everything, but I would like to select multiple fields with **jsonName** and **jsonValue** and not just one.
For example, I would like both _"url"_ and _"name"_ in jsonValue and only _"name"_ in jsonName. 
So in this way country name is shown in the dropdown and both url and country name are added to the result.
I don't know if this is possible, thank you in advance in any case ^^

**Comment by aeberhart** [comment]
Ok got it. Actually we decided against allowing this. It would mean that you could have "type": "object" and use select to determine its value.

I guess we could allow a single "select" to edit an object instead of using one subcomponent per field:

```
{
  type: object,
  widget: select,
  ...
}
```

**Comment by Salmu96** [comment]
> Ok got it. Actually we decided against allowing this. It would mean that you could have "type": "object" and use select to determine its value.
> 
> I guess we could allow a single "select" to edit an object instead of using one subcomponent per field:
> 
> ```
> {
>   type: object,
>   widget: select,
>   ...
> }
> ```

I'm okay even with this configuration: https://stackblitz.com/edit/dashjoin-qviaz6?file=src%2Fapp%2Fapp.component.ts as regard the results, but I don't understand why _"[object Object]"_ appears in the dropdown menu, I'd like there to have country names instead of  _"[object Object]"_ in the dropdown menu.

![image](https://user-images.githubusercontent.com/72612398/96889523-f7c52e00-1486-11eb-940c-4c72004a0ad6.png)

**Comment by aeberhart** [comment]
This line checks that both jsonName and jsonValue are set:
https://github.com/dashjoin/json-schema-form/blob/a0456e75db818157b54b3e4260b8e14440f88aec/projects/dashjoin/json-schema-form/src/lib/choice.ts#L137

Since this is false, we use the array items as both select display and value and pretend

```
{
      "name": "India",
      "url": "https://en.wikipedia.org/wiki/Demographics_of_India"
}
```

is a string. So you see [object Object] as the default Object toString and the actual object as the selected "string".

**Comment by Salmu96** [comment]
> This line checks that both jsonName and jsonValue are set:
> https://github.com/dashjoin/json-schema-form/blob/a0456e75db818157b54b3e4260b8e14440f88aec/projects/dashjoin/json-schema-form/src/lib/choice.ts#L137
> 
> Since this is false, we use the array items as both select display and value and pretend
> 
> ```
> {
>       "name": "India",
>       "url": "https://en.wikipedia.org/wiki/Demographics_of_India"
> }
> ```
> 
> is a string. So you see [object Object] as the default Object toString and the actual object as the selected "string".
Ok thanks a lot, now I got it. 
But in your opinion is there a way to fix that [object Object] into the "property desired" ?

**Comment by aeberhart** [comment]
We could allow jsonPointers for jsonName and jsonValue in order to handle more complex json responses.
In addition, we can allow widget: select for type object.
Changing the issue title to make this clear.
Thanks for your input!

**Comment by Salmu96** [comment]
> We could allow jsonPointers for jsonName and jsonValue in order to handle more complex json responses.
> In addition, we can allow widget: select for type object.
> Changing the issue title to make this clear.
> Thanks for your input!


Thanks a lot to you @aeberhart for your precious help! You are always very kind and helpful ^^

**Comment by aeberhart** [comment]
@Salmu96, in the meantime we introduced a new mechanism for processing response data in order to transform it to name / value choices. Check out this example:

https://dashjoin.github.io/#/example/jsonName

The schema is as follows:

```
{
  "type": "string",
  "choicesUrl": "/assets/autocomplete-complex.json",
  "jsonata": "$map(result, function($i){ {\"name\":$i.name, \"value\":$i.url} })",
  "choicesVerb": "GET"
}
```

We replaced jsonName and jsonValue with a JSONata expression (https://jsonata.org/). JSONata is ideal for transforming JSON into the desired structure. In this example, we pick the JSON field called "result" and map the objects containing url and name into the desired structure (objects with name and value).
